pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -0,0 +1,394 @@
+		}
+		srcPort = uint16(tcp.SrcPort)
+		dstPort = uint16(tcp.DstPort)
+	}
+
+	return src, srcPort, dst, dstPort, nil
+}
+
+// CreateRawTCPPacket creates a TCP packet with the specified parameters
+func CreateRawTCPPacket(sourceIP net.IP, sourcePort uint16, destIP net.IP, destPort uint16, ttl int, flags byte) (*ipv4.Header, []byte, error) {
+	ipHdr := ipv4.Header{
+		Version:  4,
+		Len:      20,
+		TTL:      ttl,
+		Protocol: 6, // TCP
+		Dst:      destIP,
+		Src:      sourceIP,
+	}
+
+	// Create TCP packet with the specified flags
+	// we'll need to vary sequence number and do
+	// some other manipulation for a paris-traceroute
+	// like accuracy but for now let's just get a trace
+	tcpPacket := make([]byte, 20)
+	binary.BigEndian.PutUint16(tcpPacket[0:2], sourcePort) // source port
+	binary.BigEndian.PutUint16(tcpPacket[2:4], destPort)   // destination port
+	binary.BigEndian.PutUint32(tcpPacket[4:8], 0)          // sequence number
+	binary.BigEndian.PutUint32(tcpPacket[8:12], 0)         // ack number
+	tcpPacket[12] = 5 << 4                                 // header length
+	tcpPacket[13] = flags
+	binary.BigEndian.PutUint16(tcpPacket[14:16], 1024) // window size
+
+	cs := tcpChecksum(&ipHdr, tcpPacket)
+	binary.BigEndian.PutUint16(tcpPacket[16:18], cs) // checksum
+
+	// TODO: calculate checksum
+
+	return &ipHdr, tcpPacket, nil
+}
+
+// MarshalPacket takes in an ipv4 header and a payload and copies
+// them into a newly allocated []byte
+func MarshalPacket(header *ipv4.Header, payload []byte) ([]byte, error) {
+	hdrBytes, err := header.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	packet := make([]byte, len(hdrBytes)+len(payload))
+	copy(packet[:len(hdrBytes)], hdrBytes)
+	copy(packet[len(hdrBytes):], payload)
+
+	return packet, nil
+}
+
+func SendPacket(rawConn *ipv4.RawConn, header *ipv4.Header, payload []byte) error {
+	if err := rawConn.WriteTo(header, payload, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checksum(data []byte) uint16 {
+	var sum uint32
+	for i := 0; i < len(data)-1; i += 2 {
+		sum += uint32(binary.BigEndian.Uint16(data[i : i+2]))
+	}
+	if len(data)%2 != 0 {
+		sum += uint32(data[len(data)-1]) << 8
+	}
+	for (sum >> 16) > 0 {
+		sum = (sum & 0xFFFF) + (sum >> 16)
+	}
+
+	return uint16(^sum)
+}
+
+func tcpChecksum(ipHdr *ipv4.Header, tcpHeader []byte) uint16 {
+	pseudoHeader := []byte{}
+	pseudoHeader = append(pseudoHeader, ipHdr.Src.To4()...)
+	pseudoHeader = append(pseudoHeader, ipHdr.Dst.To4()...)
+	pseudoHeader = append(pseudoHeader, 0) // reserved
+	pseudoHeader = append(pseudoHeader, byte(ipHdr.Protocol))
+	pseudoHeader = append(pseudoHeader, 0, byte(len(tcpHeader))) // tcp length
+
+	return checksum(append(pseudoHeader, tcpHeader...))
+}
+
+func listenAnyPacket(icmpConn *ipv4.RawConn, tcpConn *ipv4.RawConn, timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16) (net.IP, uint16, layers.ICMPv4TypeCode, error) {
```

Comment:
two nits: 
1. documentation on this is appreciated! It's hard to tell what's happening just from the type signature  
2. I recommend reordering this file in function call order so this method appears before the helpers used by it https://github.com/uber-go/guide/blob/master/style.md#function-grouping-and-ordering